File: internal/cli/cmd/push.go
package cmd

import (
	"fmt"

	"github.com/spf13/cobra"

	"github.com/archivekeep/archivekeep/internal/cli/currentarchive"
	"github.com/archivekeep/archivekeep/x/operations/sync"
)

func pushCmd() *cobra.Command {
	var flags struct {
		resolveMoves             bool
		enableDuplicateReduction bool
		enableDuplicateIncrease  bool

		additiveDuplicating bool
	}

	var connectionFlags ConnectionFlags

	cmdFunc := func(cmd *cobra.Command, args []string) error {
		currentArchive, err := currentarchive.Open()
		if err != nil {
			return fmt.Errorf("open current archive: %w", err)
		}
		remoteArchive, err := openOtherArchive(cmd, args[0], currentArchive, connectionFlags)
		if err != nil {
			return fmt.Errorf("open remote archive: %w", err)
		}

		return performSync(
			cmd,
			sync.Options{
				ResolveMoves:             flags.resolveMoves,
				EnableDuplicateIncrease:  flags.enableDuplicateIncrease,
				EnableDuplicateReduction: flags.enableDuplicateReduction,

				AdditiveDuplicating: flags.additiveDuplicating,
			},
			"current", "remote", "push",
			currentArchive, remoteArchive,
		)
	}

	cmd := &cobra.Command{
		Use:   "push [other archive location]",
		Short: "Pushes changes from current archive to remote archive",
		Args:  cobra.ExactArgs(1),
		RunE:  cmdFunc,
	}

	connectionFlags.register(cmd)

	cmd.Flags().BoolVar(&flags.resolveMoves, "resolve-moves", false, "resolves added files against checksums of existing remote files")
	cmd.Flags().BoolVar(&flags.enableDuplicateIncrease, "allow-duplicate-increase", false, "allow to increase duplication of existing files")
	cmd.Flags().BoolVar(&flags.enableDuplicateReduction, "allow-duplicate-reduction", false, "allow to decrease duplication of existing files")

	cmd.Flags().BoolVar(&flags.additiveDuplicating, "additive-duplicating", false, "push all new local files to remote without resolving moves or checking for duplication increase")

	return cmd
}

File: src/main/kotlin/org/archivekeep/cli/commands/Push.kt
package org.archivekeep.cli.commands

import org.archivekeep.cli.MainCommand
import org.archivekeep.core.operations.*
import picocli.CommandLine.*
import picocli.CommandLine.Model.CommandSpec
import java.io.PrintWriter
import java.util.concurrent.Callable

@Command(
    name = "push",
    description = ["Pushes changes from current archive to other repository."],
)
class Push : Callable<Int> {
    @Spec
    lateinit var spec: CommandSpec

    @ParentCommand
    private lateinit var mainCommand: MainCommand

    @Parameters(
        paramLabel = "other",
        description = ["Other repository location"],
    )
    private lateinit var otherArchiveLocation: String

    @Option(
        names = ["--resolve-moves"],
        description = ["resolves added files against checksums of existing remote files"]
    )
    var resolveMoves: Boolean = false

    @Option(
        names = ["--allow-duplicate-increase"],
        description = ["allow to increase duplication of existing files"]
    )
    var allowDuplicateIncrease: Boolean = false

    @Option(
        names = ["--allow-duplicate-reduction"],
        description = ["allow to decrease duplication of existing files"]
    )
    var allowDuplicateReduction: Boolean = false

    @Option(
        names = ["--additive-duplicating"],
        description = ["push all new local files to remote without resolving moves or checking for duplication increase"]
    )
    var additiveDuplicating: Boolean = false

    val out: PrintWriter
        get() = spec.commandLine().out

    override fun call(): Int {
        val currentArchive = mainCommand.openCurrentArchive()
        val otherArchive = mainCommand.openOtherArchive(otherArchiveLocation)

        val syncMode = when {
            additiveDuplicating -> RelocationSyncMode.AdditiveDuplicating
            resolveMoves -> RelocationSyncMode.Move(
                allowDuplicateIncrease = allowDuplicateIncrease,
                allowDuplicateReduction = allowDuplicateReduction
            )

            else -> RelocationSyncMode.Disabled
        }


        val preparedSync = try {
            SyncOperation(syncMode).prepare(currentArchive.repo, otherArchive)
        } catch (e: RelocationsPresentButDisabledException) {
            out.println(e.message)
            out.println("Enable relocations with --resolve-moves, or switch to --additive-duplicating mode")
            return 1
        } catch (e: DuplicationIncreasePresentButDisabledException) {
            out.println(e.message)
            out.println("Enable duplication increase with --allow-duplicate-increase, or switch to --additive-duplicating mode")
            return 1
        }

        preparedSync.execute(
            currentArchive.repo,
            otherArchive,
            prompter = { step ->
                when (step) {
                    is AdditiveRelocationsSyncStep ->
                        mainCommand.askForConfirmation("Do you want to perform additive duplicating?")

                    is RelocationsMoveApplySyncStep ->
                        mainCommand.askForConfirmation("Do you want to perform moves?")

                    is NewFilesSyncStep ->
                        mainCommand.askForConfirmation("Do you want to push new files?")
                }
            },
            logger = object : SyncLogger {
                override fun onFileStored(filename: String) {
                    out.println("file stored: $filename")
                }

                override fun onFileMoved(from: String, to: String) {
                    out.println("file moved: $from -> $to")
                }
            }
        )

        return 0
    }
}
